From a4a1b9b0283169b92209cae8230d0ace74d6a99f Mon Sep 17 00:00:00 2001
From: Piyush Gupta <gpiyush@vmware.com>
Date: Fri, 8 Apr 2022 07:18:22 +0000
Subject: [PATCH 07/16] m_{postinstall,preinstall}.py: Fix post install
 arbitrary script execution

Post and pre install scripts are listed using os.listdir() after copying
to temp dir which doesn't confirm the order of execution of scripts
as provided in kickstart.

Current change maintains a seperate list of scripts with their order
of execution and checks if the script is executable before running it.

If not executable script execution is skipped.

Change-Id: Ia6312330ce919f389994e6ffc06bd87de8a9ffaa
---
 photon_installer/modules/m_postinstall.py | 8 +++++++-
 photon_installer/modules/m_preinstall.py  | 8 +++++++-
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/photon_installer/modules/m_postinstall.py b/photon_installer/modules/m_postinstall.py
index 6e1dbd4..c45bba4 100644
--- a/photon_installer/modules/m_postinstall.py
+++ b/photon_installer/modules/m_postinstall.py
@@ -17,6 +17,7 @@ def execute(installer):
 
     tempdir = "/tmp/tempscripts"
     tempdir_full = installer.photon_root + tempdir
+    scripts = []
     if not os.path.exists(tempdir_full):
         os.mkdir(tempdir_full)
 
@@ -30,13 +31,18 @@ def execute(installer):
         with open(script_file, 'wb') as outfile:
             outfile.write("\n".join(script).encode())
         os.chmod(script_file, 0o700)
+        scripts.append('builtin_postinstall.sh')
 
     if 'postinstallscripts' in installer.install_config:
         for scriptname in installer.install_config['postinstallscripts']:
             script_file = installer.getfile(scriptname)
             shutil.copy(script_file, tempdir_full)
+            scripts.append(os.path.basename(scriptname))
 
-    for script in os.listdir(tempdir_full):
+    for script in scripts:
+        if not os.access(os.path.join(tempdir_full, script), os.X_OK):
+            installer.logger.warning("Post install script {} is not executable. Skipping execution of script.".format(script))
+            continue
         installer.logger.info("Running script {}".format(script))
         installer.cmd.run_in_chroot(installer.photon_root, "{}/{}".format(tempdir, script))
 
diff --git a/photon_installer/modules/m_preinstall.py b/photon_installer/modules/m_preinstall.py
index 3f08db5..ba1ccbb 100644
--- a/photon_installer/modules/m_preinstall.py
+++ b/photon_installer/modules/m_preinstall.py
@@ -16,6 +16,7 @@ def execute(installer):
         return
 
     tempdir = "/tmp/tempscripts"
+    scripts = []
     if not os.path.exists(tempdir):
         os.mkdir(tempdir)
 
@@ -28,13 +29,18 @@ def execute(installer):
         with open(script_file, 'wb') as outfile:
             outfile.write("\n".join(script).encode())
         os.chmod(script_file, 0o700)
+        scripts.append('builtin_preinstall.sh')
 
     if 'preinstallscripts' in installer.install_config:
         for scriptname in installer.install_config['preinstallscripts']:
             script_file = installer.getfile(scriptname)
             shutil.copy(script_file, tempdir)
+            scripts.append(os.path.basename(scriptname))
 
-    for script in os.listdir(tempdir):
+    for script in scripts:
+        if not os.access(os.path.join(tempdir, script), os.X_OK):
+            installer.logger.warning("Post install script {} is not executable. Skipping execution of script.".format(script))
+            continue
         installer.logger.info("Running script {}".format(script))
         cmd = ["/bin/bash"]
         cmd.append("-c")
-- 
2.23.1

